src/middleware/TestingHandler.ts

```diff
@@ -0,0 +1,255 @@
+	 * @param {string} urlMethod - the complete URL
+	 * @returns the relative URL
+	 */
+	private getRelativeURL(pattern: RegExp, urlMethod: string): string {
+		// just helps in getting the relative URL from the complete url, just to match the url as in manual map
+		urlMethod = urlMethod.replace(pattern, "");
+		return urlMethod;
+	}
+
+	/**
+	 * @private
+	 * To fetch the status code from the map(if needed), then returns response by calling createResponse
+	 * @param {TestingHandlerOptions} testingHandlerOptions - The TestingHandlerOptions object
+	 * @param {string} requestURL - the URL for the request
+	 * @param {string} requestMethod - the API method for the request
+	 * @returns Response object
+	 */
+	private setStatusCode(testingHandlerOptions: TestingHandlerOptions, requestURL: string, requestMethod: string): Response {
+		try {
+			testingHandlerOptions.statusMessage = "Some error happened";
+			if (testingHandlerOptions.testingStrategy === TestingStrategy.MANUAL) {
+				if (testingHandlerOptions.statusCode === undefined) {
+					try {
+						const urlMethod = this.getRelativeURL(new RegExp("http(s)://graph.microsoft.com/[^/]*", "g"), requestURL);
+						testingHandlerOptions.statusCode = this.manualMap.get(urlMethod).get(requestMethod);
+					} catch (error) {
+						const urlMethod = this.getRelativeURL(new RegExp("http(s)://graph.microsoft.com/[^/]*", "g"), requestURL);
+						this.manualMap.forEach((value: Map<string, number>, key: string) => {
+							const regexUrl = new RegExp(key);
+							if (regexUrl.test(urlMethod)) {
+								testingHandlerOptions.statusCode = this.manualMap.get(key).get(requestMethod);
+							}
+						});
+
+						if (testingHandlerOptions.statusCode === undefined) {
+							if (requestURL === this.redirectURL) {
+								testingHandlerOptions.statusCode = 404;
+							} else {
+								throw new Error("API not available in map");
+							}
+						}
+					}
+				} else {
+					if (requestURL === this.redirectURL) {
+						const statusCode: number = testingHandlerOptions.statusCode;
+						if (statusCode === 301 || statusCode === 302 || statusCode === 303 || statusCode === 307 || statusCode === 308) {
+							testingHandlerOptions.statusCode = 404;
+						}
+					}
+				}
+			} else if (testingHandlerOptions.testingStrategy === TestingStrategy.RANDOM) {
+				testingHandlerOptions.statusCode = this.getStatusCode(requestMethod);
+			}
+
+			return this.createResponse(testingHandlerOptions, requestURL);
+		} catch (error) {
+			throw error;
+		}
+	}
+
+	/**
+	 * @private
+	 * To get the options for execution of the middleware
+	 * @param {Context} context - The context object
+	 * @returns A options for middleware execution
+	 */
+	private getOptions(context: Context): TestingHandlerOptions {
+		let options: TestingHandlerOptions;
+		if (context.middlewareControl instanceof MiddlewareControl) {
+			options = context.middlewareControl.getMiddlewareOptions(TestingHandlerOptions) as TestingHandlerOptions;
```

Comment:
Add null checks for the input parameter & middleware control

Reply:
We have set Default cases and some error handling, so it seems we won't need null checks here.